core/src/saros/versioning/VersionManager.java

```diff
@@ -27,188 +32,144 @@
+      packet -> {
+        log.debug("received version info from " + packet.getFrom());
+        handleVersionInfo(packet);
+      };
 
-  private final Version localVersion;
-  private final ITransmitter transmitter;
-  private final IReceiver receiver;
-
-  private final PacketListener versionRequestListener =
-      new PacketListener() {
-        /*
-         * As some of the logic may changed in the next Saros versions this
-         * method MUST be robust as possible, expect weird and faulty data !
-         */
-        @Override
-        public void processPacket(Packet packet) {
-
-          LOG.debug("received version request from " + packet.getFrom());
-
-          VersionExchangeExtension versionExchangeRequest =
-              VersionExchangeExtension.PROVIDER.getPayload(packet);
-
-          if (versionExchangeRequest == null) {
-            LOG.warn("cannot reply to version request, packet is malformed");
-            return;
-          }
-
-          VersionExchangeExtension versionExchangeResponse = new VersionExchangeExtension();
-
-          createResponseData:
-          {
-            versionExchangeResponse.set(VERSION_KEY, localVersion.toString());
-
-            String remoteVersionString = versionExchangeRequest.get(VERSION_KEY);
-            if (remoteVersionString == null) {
-              LOG.warn("remote version string not found in version exchange data");
-              break createResponseData;
-            }
-
-            Version remoteVersion = Version.parseVersion(remoteVersionString);
-            if (remoteVersion == Version.INVALID) {
-              LOG.warn("remote version string is invalid: " + remoteVersionString);
-              break createResponseData;
-            }
-
-            versionExchangeResponse.set(ID_KEY, versionExchangeRequest.get(ID_KEY));
-          }
-
-          IQ reply = VersionExchangeExtension.PROVIDER.createIQ(versionExchangeResponse);
-          reply.setType(IQ.Type.RESULT);
-          reply.setTo(packet.getFrom());
-
-          try {
-            transmitter.sendPacket(reply);
-          } catch (IOException e) {
-            LOG.error("could not send version response to " + packet.getFrom(), e);
-          }
-
-          LOG.debug("send version response to " + packet.getFrom());
+  private final IContactsUpdate contactsUpdateListener =
+      (contact, updateType) -> {
+        if (updateType == UpdateType.STATUS) {
+          if (contact.map(c -> c.getStatus().isOnline()).orElse(false))
+            sendVersionInfo(contact.get());
+          else contact.ifPresent(remoteInfo::remove);
+        } else if (updateType == UpdateType.REMOVED) {
+          contact.ifPresent(remoteInfo::remove);
+        } else if (updateType == UpdateType.NOT_CONNECTED) {
+          remoteInfo.clear();
         }
       };
 
   public VersionManager(
-      @SarosVersion String version, final IReceiver receiver, final ITransmitter transmitter) {
-
+      @SarosVersion String version,
```

Comment:
As @m273d15 already said, you are mixing concerns. This no longer just a VersionManager. In fact the whole versionlogic is now more or less a one-liner in that class.

Reply:
As I have written: >Not included now for readability: would rename VersionManager to InfoManager and the package saros.versioning to saros.communication.info

I added a Commit renaming the Class and package.